contrib/platform/test/com/sun/jna/platform/win32/SAFEARRAYTest.java

```diff
@@ -665,10 +667,13 @@ public void testADODB() {
         SAFEARRAY arr = SAFEARRAY.createSafeArray(1);
         arr.putElement(new VARIANT("System.ItemUrl"), 0);
         boolean exceptionCaught = false;
+        VARIANT itemName = new VARIANT("System.ItemName");
         try {
-            arr.putElement(new VARIANT("System.ItemName"), 1);
+            arr.putElement(itemName, 1);
```

Comment:
I would simplify this to (string is special cased):

```suggestion
            arr.putElement("System.ItemName", 1);
```

and remove the code in the new lines 670+675+676

https://github.com/java-native-access/jna/blob/25df525d4cd8c8b96b52866d3e869c6ba2ccfe4c/contrib/platform/src/com/sun/jna/platform/win32/OaIdl.java#L709-L714

Reply:
OK, but that does prompt some follow-on questions:
1. Should I also change the `putElement()` that follows (line 680 new/675 original) after redimensioning the array to match (in order to test this special branch of the code in both success and failure)?
2. I'm struggling to understand how line 712 in `Oaldl.java` is correct.  We are freeing native memory and only saving the pointer in the array.  Wouldn't later access of this array member risk an invalid memory access error?  (I suppose changing (1) as I just suggested would actually test this.)

Reply:
>     1. Should I also change the `putElement()` that follows (line 680 new/675 original) after redimensioning the array to match (in order to test this special branch of the code in both success and failure)?

To my understanding, yes that usage also should be updated.

>     2. I'm struggling to understand how line 712 in `Oaldl.java` is correct.  We are freeing native memory and only saving the pointer in the array.  Wouldn't later access of this array member risk an invalid memory access error?  (I suppose changing (1) as I just suggested would actually test this.)

This assumes, that `SafeArrayPutElement` just assigns a pointer (and does some locking). I would be surprised if it would the case. And in fact the [remarks](https://docs.microsoft.com/en-us/windows/win32/api/oleauto/nf-oleauto-safearrayputelement) say:

> [...] If the data element is a string, object, or variant, the function copies it correctly when the safe array is destroyed. If the existing element is a string, object, or variant, it is cleared correctly. If the data element is a VT_DISPATCH or VT_UNKNOWN, AddRef is called to increment the object's reference count. [...]

The wording is a bit strange, but at least the `Unknown`/`Dispatch` case is explicitly called out to increment the reference count, it would be strange if the BSTR would not be handled identically. The situation is different, when you directly access the backing memory (which is possible).

These references at least support that thesis:

- https://microsoft.public.vc.atl.narkive.com/hr6Ad3rs/safearray-of-bstr-help-desperatly-needed#post2
- https://stackoverflow.com/a/8807892/9415272

Reply:
OK, so I've read enough to know that COM figures it all out, probably copying the data in BSTR (when successful).

However, the code snippet above starting at line 709 assumes a type of `VT_BSTR` but as written code drops to the next switch, `VT_VARIANT`.  We would need to specify the type of the array as BSTR for the simplified String allocation to work:
```
SAFEARRAY arr = SAFEARRAY.createSafeArray(new WTypes.VARTYPE(Variant.VT_BSTR), 1);
```

Reply:
And I think that needs redefining of the tests types (e.g., `columnList.setValue(Variant.VT_ARRAY | Variant.VT_VARIANT, arr);` should be an array of `BSTR`?  And `assertThat(sa.getVarType().intValue(), is(Variant.VT_VARIANT));` should test for `BSTR` type?   And changing the casting when fetching the results?

Reply:
A f*** - sorry, yes we are not seeing a `BSTR` array, but a `VARIANT` array. Your original fix was right.

Reply:
We could probably allow special handling of the String in the VARIANT case, wrapping a BSTR there. :)

But for the next release I'll merge this as is.